neqo-crypto/src/aead.rs

```diff
@@ -13,13 +13,75 @@ use std::{
 
 use crate::{
     constants::{Cipher, Version},
-    err::Res,
+    err::{sec::SEC_ERROR_BAD_DATA, Error, Res},
     experimental_api,
     p11::{PK11SymKey, SymKey},
     scoped_ptr,
     ssl::{PRUint16, PRUint64, PRUint8, SSLAeadContext},
 };
 
+/// Trait for AEAD (Authenticated Encryption with Associated Data) operations.
+///
+/// This trait provides a common interface for both real and null AEAD implementations,
+/// eliminating code duplication and allowing for consistent usage patterns.
+pub trait Aead {
+    /// Create a new AEAD instance.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error` when the underlying crypto operations fail.
+    fn new(version: Version, cipher: Cipher, secret: &SymKey, prefix: &str) -> Res<Self>
+    where
+        Self: Sized;
+
+    /// Get the expansion size (authentication tag length) for this AEAD.
+    fn expansion(&self) -> usize
+    where
+        Self: Sized;
```

Comment:
Does this trait need to be [dyn compatible](https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility)? If not, would the below work? Would simplify the usage sites.

```suggestion
    /// Get the expansion size (authentication tag length) for this AEAD.
    fn expansion() -> usize
    where
        Self: Sized;
```

Reply:
You don't need the Sized constraint on this function; it takes `&self`.

```suggestion
    /// Get the expansion size (authentication tag length) for this AEAD.
    fn expansion(&self) -> usize
```

The reason you need `&self` is that the cipher suite (which is a parameter of `Aead::new`) determines the expansion.